internal/cache/metadata/type_cache.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/googlecloudplatform/gcsfuse/internal/util"
+)
+
+type Type int
+
+const (
+	UnknownType     Type = 0
+	SymlinkType     Type = 1
+	RegularFileType Type = 2
+	ExplicitDirType Type = 3
+	ImplicitDirType Type = 4
+	NonexistentType Type = 5
+)
+
+// TypeCache is a (name -> Type) map.
+// It maintains TTL for each entry for supporting
+// TTL-based expiration.
+// Sample usage:
+//
+//	tc := NewTypeCache(size, ttl)
+//	tc.Insert(time.Now(), "file", RegularFileType)
+//	tc.Insert(time.Now(), "dir", ExplicitDirType)
+//	tc.Get(time.Now(),"file") -> RegularFileType
+//	tc.Get(time.Now(),"dir") -> ExplicitDirType
+//	tc.Get(time.Now()+ttl+1ns, "file") -> internally tc.Erase("file") -> UnknownType
+//	tc.Erase("dir")
+//	tc.Get(time.Now(),"dir") -> UnknownType
+type TypeCache interface {
+	// Insert inserts the given entry (name -> type)
+	// with the entry-expiration at now+ttl.
+	Insert(now time.Time, name string, it Type)
+	// Erase removes the entry with the given name.
+	Erase(name string)
+	// Get returns the entry with given name, and also
+	// records this entry as latest accessed in the cache.
+	// If now > expiration, then entry is removed from cache, and
+	// UnknownType is returned.
+	// If entry doesn't exist in the cache, then
+	// UnknownType is returned.
+	Get(now time.Time, name string) Type
+}
+
+type cacheEntry struct {
+	expiry    time.Time
+	inodeType Type
+}
+
+func (ce cacheEntry) Size() uint64 {
+	return uint64(unsafe.Sizeof(ce))
+}
+
+// A cache that maps from a name to information about the type of the object
+// with that name. Each name N is in one of the following states:
+//
+//   - Nothing is known about N.
+//   - We have recorded that N is a file.
+//   - We have recorded that N is a directory.
+//   - We have recorded that N is both a file and a directory.
+//
+// Must be created with NewTypeCache. May be contained in a larger struct.
+// External synchronization is required.
+type typeCache struct {
+	/////////////////////////
+	// Constant data
+	/////////////////////////
+
+	ttl time.Duration
+
+	/////////////////////////
+	// Mutable state
+	/////////////////////////
+
+	// A cache mapping names to the cache entry.
+	// Each cache entry (including negative ones) is a fixed 32-byte in size (especially for 64-bit linux machines, may differ on other platforms).
+	//
+	// INVARIANT: entries.CheckInvariants() does not panic
+	// INVARIANT: Each value is of type cacheEntry
+	entries *lru.Cache
+}
+
+// NewTypeCache creates an LRU-policy-based cache with given max-size and TTL.
+// Any entry whose TTL has expired, is removed from the cache on next access (Get).
+// When insertion of next entry would cause size of cache > sizeInMB,
+// older entries are evicted according to the LRU-policy.
+// If either of TTL or sizeInMB is zero, nothing is ever cached.
+func NewTypeCache(sizeInMB int, ttl time.Duration) TypeCache {
+	if ttl > 0 && sizeInMB != 0 {
+		if sizeInMB < -1 {
```

Comment:
This is a public function, so ideally it should check any inputs sent to it (especially when it can check them so easily), and not assume who/what the caller is.

Anyway, I have removed the check now.